Configurable exclusion of labels in User Operator (#60)
strimzi/proposals: 039-user-operator-custom-filter-for-labels-.md

# User Operator: Configurable exclusion of labels

This proposal is about adding the ability to filter the automatically assigned labels.
When User Operator creates `KafkaUser`, it also creates an associated Secret, where it automatically adds the label
`app.kubernetes.io/instance: <username>`.
Such a label could be used for many uses, and different users have various requirements and expectations.

> Nevertheless, it could also lead to not likeable scenarios (e.g., [repeated deletion and re-creation of specific Secret](https://github.com/strimzi/strimzi-kafka-operator/issues/5690)).

Reviewer: ```suggestion
Nevertheless, it could also lead to undesirable scenarios (e.g., [repeated deletion and re-creation of specific Secret](https://github.com/strimzi/strimzi-kafka-operator/issues/5690)).
```

Therefore, we propose to implement configurable exclusion of labels.

## Current situation

> Presently, when a user attempts to create a KafkaUser defined as follows:

Reviewer: ```suggestion
Presently, when a user attempts to create a `KafkaUser` defined as follows:
```

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaUser
metadata:
  name: my-user
  labels:
    strimzi.io/cluster: my-cluster
spec:
  authentication:
    type: tls
```

> The User Operator then creates an associated Kubernetes Secret.

Reviewer: ```suggestion
The User Operator then creates an associated Kubernetes `Secret`.
```

```yaml
kind: Secret
metadata:
  labels:
    app.kubernetes.io/instance: my-user
    app.kubernetes.io/managed-by: strimzi-user-operator
    app.kubernetes.io/name: strimzi-user-operator
    app.kubernetes.io/part-of: strimzi-my-user
    strimzi.io/cluster: my-cluster
    strimzi.io/kind: KafkaUser
  name: my-user
  namespace: myproject
...
```
We can see that User Operator automatically adds the `app.kubernetes.io/instance: my-user` label.

> In some situations, we do not want a such label to be automatically created (i.e., [endless deletion and creation of specific Secret](https://github.com/strimzi/strimzi-kafka-operator/issues/5690))
> and therefore, one way how to get rid of this already assigned label is to filter it.

Reviewer: ```suggestion
If we do not want the label that has been assigned, the only way to get rid of it is to filter it.
```

Reviewer: I made the above suggestion because we have the same link in the intro.

## Proposal

This proposal suggests adding configurable exclusion of labels.

> We can implement this feature similarly to [issue](https://github.com/strimzi/strimzi-kafka-operator/pull/4791).

Reviewer: ```suggestion
We can implement this feature similarly to [issue 4394](https://github.com/strimzi/strimzi-kafka-operator/pull/4791).
```

> We can create an environment variable and inject such value through the `Kafka` custom resource.

Reviewer: ```suggestion
We can create an environment variable and inject the value through the `Kafka` custom resource.
```

> Specifically, in the `spec.entityOperator.template.topicOperatorContainer.env`.

Reviewer 2: you meant `userOperatorContainer` here I guess

For instance, we can have the following `Kafka` resource:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  ...
  entityOperator:
    template:
      userOperatorContainer:
        env:
          name:   STRIMZI_LABELS_EXCLUSION_PATTERN
          value:  "^app.kubernetes.io/.*$"
    topicOperator: {}
    userOperator: {}
```
Then, we would obtain a value from the environment variable and parse it in the `UserOperatorConfig` class.

> Based on the environment variable value, we will determine what to do (a) env is not specified, and therefore we do not exclude any label from Secret
> (b) env is specified by a regex, and then we will filter and remove such labels which match the input regex.

Reviewer: ```suggestion
Based on the environment variable value, (a) when `env` is not specified, we do not exclude any label from the secret
(b) when `env` is specified by a regex, we filter and remove those labels which match the regex.
```

### More implementation details

> ### 1 Parsing part

Author: fixed

As we mentioned details about the `UserOperatorConfig` class, we will elaborate more on the specific details in this section.
Moreover, we will obtain the environment variable inside the `fromMap` method used to construct the class.
```java
public static UserOperatorConfig fromMap(Map<String, String> map) {
    ...
    // 
    String strimziLabelsExclusionPatternEnvVar = map.get(UserOperatorConfig.STRIMZI_LABELS_EXCLUSION_PATTERN);
    
    if (strimziLabelsExclusionPattern != null) {
        // note that we will compile such regex into FSM only once here and thus eliminate workload inside KafkaUserModel
        strimziLabelsExclusionPattern = Pattern.compile(strimziLabelsExclusionPatternEnvVar);
    }
    ...
}
```

#### Exclusion part

The exclusion part will be placed in the `KafkaUserModel` class.
Moreover, we would need to get the value of the environment variable to such a class.

> In the `UserOperatorConfig` class, we will obtain value, and then in the `KafkaUserOperaor` class, specifically in the following method:

Reviewer: ```suggestion
In the `UserOperatorConfig` class, we will obtain the environment variable value, and then in the `KafkaUserOperator` class, specifically in the following method:
```

Reviewer: This bit seems unfinished _and then in the `KafkaUserOperaor` class_...?

protected Future<KafkaUserStatus> createOrUpdate(Reconciliation reconciliation, KafkaUser resource) {
    KafkaUserModel user;
    KafkaUserStatus userStatus = new KafkaUserStatus();

    try {
        user = KafkaUserModel.fromCrd(resource, config.getSecretPrefix(), config.isAclsAdminApiSupported(),config.isKraftEnabled(), 

>             config.getStrimziLabelsExclusionPattern); // <-- this one we will inject into KafkaUserModel class

Reviewer 3: This way, you will ahve to compile the variable again and again in every reocnciliation. That seems like waste of resources.

Author: I have changed the String type of strimziLabelsExclusionPattern to the `Pattern` and also compile such regex in the parsing part (i.e., in the UserOperatorConfig), which should eliminate this workload...

            config.getStrimziLabelsExclusionPattern); // <-- this one we will inject into KafkaUserModel class
    } catch (Exception e) {
        StatusUtils.setStatusConditionAndObservedGeneration(resource, userStatus, Future.failedFuture(e));
        return Future.failedFuture(new ReconciliationException(userStatus, e));
    }
    ...
}
```

> Then, we can use such value to store it as an instance attribute of `KafkaUserModel` and implement pre-processing of labels in the

Reviewer: ```suggestion
Then, we can use the environment variable value to store it as an instance attribute of `KafkaUserModel` and implement pre-processing of labels in the
```

```java
protected Secret createSecret(Map<String, String> data) {
    final Secret secret = return new SecretBuilder()
        .withNewMetadata()
            .withName(getSecretName())
            .withNamespace(namespace)
            .withLabels(Util.mergeLabelsOrAnnotations(labels.toMap(), templateSecretLabels))
            .withAnnotations(Util.mergeLabelsOrAnnotations(null, templateSecretAnnotations))
            .withOwnerReferences(createOwnerReference())
        .endMetadata()
        .withType("Opaque")
        .withData(data)
        .build();
    
    // here, we have to do post-processing (i.e., filtering) of labels by exclusion pattern

>     // filter by value of instance variable `this.getStrimziLabelsExclusionPattern`

Author: Well, it could be done here or before the creation of an instance. But I think pre-processing would be better in this case. I have changed the order :)

    
    return secret;
}
```

## Compatibility

This proposal does not change any of the existing CRDs or the Kubernetes secrets that are being created.

> The user only needs to modify Kafka's custom resource to use such a feature.

Reviewer 3: Setting it through Kafka CR is only one way. You can use also standalone User Operator. So you should probably not focus too much on it.

Reviewer 3: This should still be addressed @see-quick

Author: So what about something like:
```
The user only needs to modify either User Operator deployment in the case of a standalone approach or Kafka's custom resource to use such a feature.
```

Reviewer 3: For example ... make it clear that this is something available and configurable in both.
